llama : tokenizer unicode codepoint categories

jaime-m-p opened the pull request:
Add all unicode [categories](https://www.compart.com/en/unicode/category) to `unicode-data.cpp`.

Currently we are limited to high categories:
* C, L, M, N, P, S, Z.

This PR allows access to subcategories:
* Cn, Cc, Cf, Co, Cs, Ll, Lm, Lo, Lt, Lu, Mc, Me, Mn, Nd, Nl, No, Pc, Pd, Pe, Pf, Pi, Po, Ps, Sc, Sk, Sm, So, Zl, Zp, Zs.

Related PR: https://github.com/ggerganov/llama.cpp/pull/8579, regex using Lu, Lt, Lm, Lo, etc.

**TODO**: Add more comments to explain the unicode regex *collapse trick* for all subcategories.

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

Commit:
Update bruteforce test:

- Compare tokenizer vocab tokens.
- Bruteforce byte token generator.
- Find minimal mismatched substring.

Commit:
Store all unicode codepoint categories

Commit:
Reimplement 'codepoint_flags' as 'codepoint_categ'

Commit:
Update unicode data

Commit:
Decode unicode data categories

Commit:
Replace 'codepoint_flags' with 'codepoint_categ'

compilade commented:
This should also help fix (at least part of) [Falcon's tokenization](https://huggingface.co/tiiuae/falcon-7b/blob/main/tokenizer.json), because the `Punctuation` pre-tokenizer type uses the `Po` category and not the broader `P` one.

(ref: <https://github.com/huggingface/tokenizers/blob/4ea2f235b0430f5db09f867b65306d6c0a5ec7ed/tokenizers/src/pre_tokenizers/punctuation.rs#L8>, which uses Rust's [`is_ascii_punctuation`](https://doc.rust-lang.org/std/primitive.char.html#method.is_ascii_punctuation) and [`is_punctuation`](https://doc.rust-lang.org/nightly/nightly-rustc/rustfmt_nightly/string/fn.is_punctuation.html))

ggerganov reviewed (approved):
> TODO: Implement unicode regex collapse trick for all subcategories.

Do you expect any problems with this? Probably we will run out of ASCII characters for the `k_ucat_cpt` map:

https://github.com/ggerganov/llama.cpp/blob/50e05353e88d50b644688caa91f5955e8bdb9eb9/src/unicode.cpp#L647-L651

Though we could dynamically generate the map based only on the used subcategories in the current regex

ggerganov commented:
The `src/llama.cpp` conflict should be easy to resolve - just accept the new `src/llama.cpp` and apply the same changes to `src/llama-vocab.cpp` instead

Commit:
Update unicode data: sorted whitespaces

Commit:
Fix codepoint_categ return types

Commit:
Add unicode_data helper functions

Commit:
Reimplement 'collapsed' unicode categories:

- Add all unicode categories.
- Fix \s with non-ASCII problem.

Commit:
Add more comments

Commit:
Merge commit '978ba3d8' into tokenizer-codepoint-categs

ggerganov commented on src/unicode.cpp:
```diff
@@ -715,91 +722,289 @@ std::vector<std::string> unicode_regex_split(const std::string & text, const std
+                        continue;
+                    case '}':
+                    case ']':
+                        assert (false);
```
```suggestion
                        GGML_ABORT("invalid regex");
```

ggerganov commented on src/unicode.cpp:
```diff
@@ -715,91 +722,289 @@ std::vector<std::string> unicode_regex_split(const std::string & text, const std
+                        if (cpts_regex[i + 1] == '?') {  // (?: (?i: (?= (?! (?<= (?<!
+                            if (cpts_regex[i + 2] == ':') {
+                                i += 2;
+                            } else if (cpts_regex[i + 2] == 'i') {
+                                i += 3;
+                                assert (cpts_regex[i] == ':');
+                            } else {
+                                i += 2 + (cpts_regex[i + 2] == '<');
+                                assert (cpts_regex[i] == '=' || cpts_regex[i] == '!');
```
```suggestion
                            } else if (cpts_regex[i + 2] == 'i') {
                                i += 3;
                                GGML_ASSERT(cpts_regex[i] == ':');
                            } else {
                                i += 2 + (cpts_regex[i + 2] == '<');
                                GGML_ASSERT(cpts_regex[i] == '=' || cpts_regex[i] == '!');
```

ggerganov reviewed (commented):
Most of the `assert` in `unicode.cpp` should be changed to `GGML_ABORT` or `GGML_ASSERT`

Commit:
minor: remove trailing whitespaces and extra semicolons

Commit:
Use GGML_ASSERT and GGML_ABORT

Commit:
Update bruteforce test: fix pyright complaints

Commit:
Update bruteforce test:
- Faster failing text range selection.
- Show unique failing texts differences.
- Add more recent models.

Commit:
Binary constants are a C++14 feature

Commit:
Fix copy/paste wrong variable

Commit:
Fix compiler complaints

Commit:
Update bruteforce test: fix binary search

Commit:
Unicode data whitespaces as ranges

Commit:
Reimplement unicode_regex_split()

Commit:
Remove invalid assert

Commit:
Update codepoint_categ:

- Reorganize category/subcategory bits.
- Regex flags for \s \w \d.

Commit:
Reimplement unicode_regex_split():

- Using std::basic_regex.
- Custom std::ctype specialization for 32bits codepoints.
- Custom std::regex_traits specialization for 32bits codepoints.
- Implementing custom 'character class expression' for \p{Xx}.
- Single pass regex preparation.

Commit:
Original regex for 'tekken'

Commit:
Remove unused function

Commit:
Using 32bit wchar_t by default, uint32_t on Windows

Commit:
Fix previous commit

Commit:
Fix compiler complaints
